### PR TITLE
git-bash (Windows only)

### PR DESCRIPTION
**Pull request by kiwi0fruit**
Windows only Git Bash repackage (like main::git and conda-forge::git) that doesn't mess with this problem: https://github.com/conda-forge/git-feedstock/issues/53 as it only adds `.\bin\bash.bat` and `.\bin\git-bash.bat`  (git bash itself is hidden in the `.\Library\git-bash`) and `.\bin` is the last in the env PATH order on Windows so `bash` from `conda-forge::git` or `msys2::posix` packages would have priority over `bash.bat` if installed.

Reasons for this package:
* I encoutered bug in posix metapackage from msys2 channel that broke my temp folder permissions but I wasn't able to reproduce it later. So I'm hesitant to use it again.
* license issues with conda-forge::git (see [this](https://github.com/conda-forge/git-feedstock/issues/56))
* conda-forge::git doesn't pass this test: `bash -c "if [[ $(echo hello | cp /dev/stdin /dev/stdout) != 'hello' ]]; then exit 1; fi"`
* I'm going to keep pip/conda behaviour in sync so I'll later add pypi package based on this pure conda package. Because of that I cannot use posix metapackage (and bash) from msys2 channel.

Notable differences from conda-forge::git package:

* git-bash is no_link pakage that runs run git bash post-install script on the user machine. It uses post-link conda script (I hope that it won't be buggy. I encoutered failing post-link script on Windows wipes Scripts folder, conda~4.6.4).

### CODE REVIEW

**Comment by kiwi0fruit** [comment]
@conda-forge/help-python please review. Travis macOS CI doesn't matter as it's Windows only package.

**Comment by kiwi0fruit** [comment]
@msarahan please review yet another win git bash repackage (it's justified).

**Comment by kiwi0fruit** [comment]
@scopatz please review yet another win git bash repackage (it's justified).

**Comment by kiwi0fruit** [comment]
@conda-forge/help-python please review

**Comment by kiwi0fruit** [comment]
I guess the package should fix this bug [git package redistrubutes some Microsoft copyrighted files + git post install script partially fails](https://github.com/conda-forge/git-feedstock/issues/56). But if you read this befire I fixed it please comment on the package itself and give permission to merge this RP later (as this may need aproval from the core team).

**Comment by kiwi0fruit** [comment]
@conda-forge/help-python @msarahan @scopatz please review.

File: recipes/git-bash/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{% set name = "git-bash" %}
+{% set version = "2.20.1" %}
+
+package:
+  name: {{ name }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/git-for-windows/git/releases/download/v{{ version }}.windows.1/PortableGit-{{ version }}-64-bit.7z.exe   # [win64]
+  sha256: 4c0fc0c77d577f8dbb3d7711283458ba1d0ce1781cfe0d49c940abbcdf18faa0   # [win64]
```
Comment (scopatz) [inline recipes/git-bash/meta.yaml]: It doesn't seem like you are actually applying the patches.

**Comment by scopatz** [comment]
Sorry @kiwi0fruit, I didn't see this until today.  I was on holiday. In the future, it is probably best to ping @conda-forge/staged-recipes  :wink:

**Comment by kiwi0fruit** [comment]
@scopatz That's OK. Luckily :( just recently I've found a bug in the git-bash post-install script when it's run on the user machine: not only error level should be OK (it's monitored now) but also there should be no stderr messages (I'm in the process of adding an additional test). So that's not the final version.

File: recipes/git-bash/meta.yaml (hunk 1 above)
Comment (kiwi0fruit) [reply recipes/git-bash/meta.yaml]: @scopatz Patches are applied in the bld.bat

**Comment by kiwi0fruit** [comment]
@scopatz please review. It finally works as intended. There was a strange bug that the test worked in conda build but failed when manually installing the package. Now it always successful.

**Comment by kiwi0fruit** [comment]
@scopatz git-bash now exports only bash.bat and git-bash.bat to the PATH. They are simply thin wrappers like
```bat
@"%~dp0\..\Library\git-bash\bin\bash.exe" %*
```
But it would be even better if they were .exe wrappers instead of .bat. But bash.exe would have to integrate bash console into cmd console when executed. My guess it that simply starting exec won't be enough. May be we need some *tty pass through.

Do you know any simple way of creating an exe wrapper without relying on Python (as this package should be pure conda package independent on Python)?

**Comment by kiwi0fruit** [comment]
See https://github.com/git-for-windows/git/issues/2089 for details about .exe wrapper I'm planning to add. Some day...

**Comment by kiwi0fruit** [comment]
[git-wrapper.c](https://github.com/git-for-windows/MINGW-packages/blob/master/mingw-w64-git/git-wrapper.c) is a bit cryptic for me to hack right now so I go with bash.bat

Please review regardless of the. exe wrapper.

**Comment by scopatz** [comment]
@msarahan, @conda-forge/core do you have thoughts here?

**Comment by dscho** [comment]
> [git-wrapper.c](https://github.com/git-for-windows/MINGW-packages/blob/master/mingw-w64-git/git-wrapper.c) is a bit cryptic for me to hack right now so I go with bash.bat

@kiwi0fruit if you can summarize what you want it to do (so that I do not have to study this here PR, for which I sadly lack the time), I can help.

**Comment by kiwi0fruit** [comment]
@dscho Thanks for help. This PR is a repackage of the portable Git Bash for Windows for conda Python. There are some folders that are automatically added to PATH when working with Python. But unfortunately I cannot place portable Git as it is so that bash.exe is added to the PATH (because of compatibility with other conda packages reasons that who knows when would be resolved).

The important part:

This package installs git to the `.\Library\git-bash` folder. And the executable to run bash now in the `.\Library\git-bash\bin\bash.exe`. The folder that is added to the PATH is `.\bin`. So I currently use `.\bin\bash.bat` to run bash:
```bat
@"%~dp0\..\Library\git-bash\bin\bash.exe" %*
```
But it would be better to have bash.exe in the PATH instead of the bash.bat so that it would behave like bash on Unix and other conda packages that have bash.exe in the PATH.

I understand that `.\Library\git-bash\bin\bash.exe` is a wrapper that does some preparations and starts `.\Library\git-bash\usr\bin\bash.exe`. In can be said that does preparations and starts `.\..\usr\bin\bash.exe`.  So I plan to tweak  `git-wrapper.c` so that it compiles bash.exe that starts `.\..\Library\git-bash\usr\bin\bash.exe` instead. That's all.

It would be great if you can help about it.

**Comment by kiwi0fruit** [comment]
@msarahan, @conda-forge/core do you have thoughts here?

**Comment by kiwi0fruit** [comment]
@scopatz Any progress?

**Comment by scopatz** [comment]
I suppose no one objects!

**Comment by kiwi0fruit** [comment]
Thanks! :)

**Pull request merged**